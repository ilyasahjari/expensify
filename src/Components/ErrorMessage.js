import React from 'react';
import { Link } from 'react-router-dom'

const ErrorMessage = () => (
    <div>
      Error 404 - <Link to="/"> Go Home </Link>
    </div>
  )
  
export default ErrorMessage;  