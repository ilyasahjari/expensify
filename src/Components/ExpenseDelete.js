import React from 'react';

const ExpenseDelete = () => (
    <div>
      this is Expense Delete
    </div> 
 )

 export default ExpenseDelete;