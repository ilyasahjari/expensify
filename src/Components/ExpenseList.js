import React from 'react'
import { connect } from 'react-redux';
import MapExpenses from './ExpenseListItem'
import getVisibleExpenses from '../selectors/expenses'


export const ExpenseList = (props) => {
    return (
        <div>
            <h1> Showing Expenses Redux List </h1>
            <div style={{
                display: 'flex',
                flexDirection: 'row',
                justifyContent: 'space-around',
                marginTop: 5,
            }}>
                {
                    props.expenses.map((expense) => {
                        return <MapExpenses key={expense.id} {...expense} />
                    })
                }
            </div>
        </div>
    )
}

const mapsToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }
}


export default connect(mapsToProps)(ExpenseList);