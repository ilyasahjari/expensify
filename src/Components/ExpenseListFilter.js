import React from 'react'
import { connect } from 'react-redux'
import { setTextFilter, sortByAmount, sortByDate } from '../actions/filters'

const ExpenseFilter = ({ dispatch }) => {
    return (
        <div>
            <h5>
                Search : <br />
                <input type='text' onChange={(e) => {
                    dispatch(setTextFilter(e.target.value))
                }} />
            </h5>

            <h5>
                sort by : <br />
                <select onChange={(e) => {
                    if (e.target.value !== " ") {
                        if (e.target.value === 'sortbyamount')
                            dispatch(sortByAmount());
                        else if (e.target.value === 'sortbydate')
                            dispatch(sortByDate());
                    }
                }}>
                    <option value=" "> </option>
                    <option value="sortbyamount">Sort By Amount</option>
                    <option value="sortbydate">Sort By Date</option>
                </select>

            </h5>

        </div>
    )
}




export default connect()(ExpenseFilter);