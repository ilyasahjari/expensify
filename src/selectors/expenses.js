// get visible expenses 

//a test to show if the filter can see the elements existing in the expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {

    const filteredExpenses =
        expenses.filter((expense) => {
            const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
            const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
            const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

            return startDateMatch && endDateMatch && textMatch;
        });
    return filteredExpenses.sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt > b.createdAt ? 1 : -1;
        } else
            if (sortBy === 'amount') {
                return a.amount > b.amount ? -1 : 1;
            }
    });
};

