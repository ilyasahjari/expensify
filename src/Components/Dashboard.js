import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter'

const Dashboard = () => (
    <div>
      <ExpenseListFilter/>
      <ExpenseList/>
    </div>
  );
  
export default Dashboard;  

