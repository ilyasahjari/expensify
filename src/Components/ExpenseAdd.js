import React, { useState } from 'react';
import { addExpense } from '../actions/expenses';
import { connect } from 'react-redux'
import DatePicker from "react-datepicker"
import "react-datepicker/dist/react-datepicker.css";


const ExpenseAdd = (props) => {

  const [description, setDescription] = useState('');
  const [note, setNote] = useState('');
  const [amount, setAmount] = useState('');
  const [createdAt, setCreationDate] = useState(new Date());


  const handleAdd = (e) => {
    e.preventDefault();
    if (description && amount) {
      props.dispatch(addExpense({
        description,
        note,
        amount,
        createdAt: Number(createdAt)
      }));
    props.history.push('/')
    }else{
      alert ("please fill the amount or description")
    }
  };

  const onDescriptionChange = (e) => {
    const description = e.target.value;
    setDescription(description)
  }

  const onNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  }

  const onAmountChange = (e) => {
    const amount = e.target.value
    setAmount(amount)
  }



  return (
    <div>
      <form onSubmit={handleAdd}>
        Description :<br />
        <input type="text" autoFocus placeholder="put a description" onChange={onDescriptionChange} ></input><br />
        Amount :<br />
        <input type="number" placeholder="price of the product" onChange={onAmountChange} ></input><br />
        Note :<br />
        <input type="text" placeholder="add a note (optional)" onChange={onNoteChange} ></input><br />
        Date : <br />
        <DatePicker
          dateFormat="dd/MM/yyyy"
          selected={createdAt}
          onChange={date => setCreationDate(date)}
        />
        <br />
        <input type="submit" value='Add Item' />
      </form>
    </div>
  )
}



export default connect()(ExpenseAdd);  