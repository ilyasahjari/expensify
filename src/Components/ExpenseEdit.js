import React, { useState } from 'react';
import { editExpense, removeExpense } from '../actions/expenses';
import { connect } from 'react-redux'
import DatePicker from 'react-datepicker'


const ExpenseEdit = (props) => {

  const [description, setDescription] = useState(props.expense ? props.expense.description : '');
  const [note, setNote] = useState(props.expense ? props.expense.note : '');
  const [amount, setAmount] = useState(props.expense ? props.expense.amount : '');
  const [createdAt, setCreationDate] = useState(props.expense ? props.expense.createdAt : new Date());

  const handleDelete = () => {
    if (description && amount) 
      props.dispatch(removeExpense(props.expense.id))
    props.history.push('/')
  }

  const handleEdit = (e) => {
    e.preventDefault();
    if (description && amount) {
      props.dispatch(editExpense(props.expense.id, {
        description,
        note,
        amount,
        createdAt: Number(createdAt)
      }));
      props.history.push('/')
    } else {
      alert("please fill the amount or description")
      props.history.push('/')
    }
  };

  const onDescriptionChange = (e) => {
    const description = e.target.value;
    setDescription(description)
  }

  const onNoteChange = (e) => {
    const note = e.target.value;
    setNote(note);
  }

  const onAmountChange = (e) => {
    const amount = e.target.value
    setAmount(amount)
  }



  return (
    <div>
     <h2> this is the Expense Edit {props.match.params.id} </h2>
      <form onSubmit={handleEdit}>
        Description :<br />
        <input type="text" autoFocus value={description} placeholder="put a description" onChange={onDescriptionChange} ></input><br />
        Amount :<br />
        <input type="number" value={amount} placeholder="price of the product" onChange={onAmountChange} ></input><br />
        Note :<br />
        <input type="text" value={note} placeholder="add a note (optional)" onChange={onNoteChange} ></input><br />
        Date : <br />
        <DatePicker
          dateFormat="dd/MM/yyyy"
          selected={createdAt}
          showYearDropdown
          showMonthDropdown
          onChange={date => setCreationDate(date)}
        />
        <br />
        <input type="submit" value='Submit Editing' />
      </form>
      <button onClick={handleDelete} >Delete Item</button>
    </div>
  )
}

const mapToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  }
}

export default connect(mapToProps)(ExpenseEdit);  