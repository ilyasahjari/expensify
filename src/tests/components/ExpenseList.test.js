import React from "react"
import {shallow} from "enzyme"
import {ExpenseList} from "../../Components/ExpenseList"
import expenses from "../fixtures/expenses"

it("Test Expense list",()=>{
    const wrapper = shallow(<ExpenseList expenses={expenses}/>)
    expect(wrapper).toMatchSnapshot(); 
})

