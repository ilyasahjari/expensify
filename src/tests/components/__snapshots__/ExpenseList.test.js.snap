// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Expense list 1`] = `ShallowWrapper {}`;
